src/backend/distributed/connection/connection_configuration.c

```diff
@@ -280,9 +280,14 @@ GetConnParams(ConnectionHashKey *key, char ***keywords, char ***values,
 	/* first step: copy global parameters to beginning of array */
 	for (paramIndex = 0; paramIndex < ConnParams.size; paramIndex++)
 	{
-		/* copy the keyword&value pointers to the new array */
-		connKeywords[paramIndex] = ConnParams.keywords[paramIndex];
-		connValues[paramIndex] = ConnParams.values[paramIndex];
+		/*
+		 * Copy the keyword&value strings to the new array since they might
+		 * be free()-d when settings are reloaded.
+		 */
+		connKeywords[paramIndex] =
+			MemoryContextStrdup(context, ConnParams.keywords[paramIndex]);
```

Comment:
I'm a bit concerned that we are introducing / adding to an already-existing memory leak because this context is `ConnectionContext`, which is never reset.

Reply:
Hmm, that looks like a bigger problem in `GetConnParams()` because everything is allocated in `ConnectionContext`. I'll look into this.

Reply:
In master branch, if we repeatedly execute the following commands, we can reproduce memory leak:
```SQL
SELECT pg_reload_conf();
SELECT citus_total_relation_size('poc_country_1');
```

This PR, with its current form, expands the memory leak to include all fields 😄 

I think I'll keep this commit as-is and fix the problem on top of this. I'm still thinking for a simple fix.

Reply:
I'd like to give an update on this issue.

First of all, `citus_relation_size()` has a completely unrelated memory leak, which confused me a lot. Here(#2618) is the fix for it.

Initially, I thought that the logic for `AddConnParam()` and/or `ResetConnParams()` in combination with SIGHUP/pg_reload_conf() is broken. However, that logic seems sane now. I was doing my tests with `citus_relation_size()` and as I mentioned above, that confused my on this topic.

Note that the fix in this PR could be seen as a completely separate issue. As I mentioned above, this fix expands the memory leak, but at least makes all elements allocated in the same context. Another fix on top of this should fix the memory leak.

The actual memory leak is happening on `ConnParamsHash` invalidation, in the enterprise repo. We keep allocating connection keywords and values in `ConnectionContext` and never deallocate them. Thus, any invalidation on `ConnParamsHash` makes the leak visible.

On citus-enterprise, if you open two sessions, and execute the followings, you'd see the leak:
```SQL
-- session 1
SELECT count(*) FROM table;
\watch 0.001
```

```SQL
-- session 2
UPDATE pg_dist_poolinfo SET poolinfo = '';
\watch 0.001
```
You'll soon see that session 1 uses a lot of memory, and keeps increasing.

Well, the fix does not seem trivial. We probably need a relatively large refactoring. My current suggestions:

- Per `MultiConnection` keep `connKeywords` and `connValues` in a per connection context (no such context exists yet). Make sure to use those variables for starting the pg connection. While `free` ing the `MultiConnetion`, delete this internal  memory context as well. (Probably a wrapper function doing both.)
- Whenever an invalidation comes, mark the entries invalid. Later, at the beginning or end of `StartConnectionEstablishment()` get rid of the invalid entries. Note that invalidation might come during `StartConnectionEstablishment()` so do not forget and break things further :) 

Or, we could allow memory leaks, but only when the values actually change by checking the values in the callbacks. Though, I don't quite like that idea.

@jasonmp85 did you assigned yourself for the review or do you want to take-over the fixes? Both works fine for me.

Reply:
I assigned myself to look over it because I was the author and I remember this code being somewhat delicately and intentionally written. It was more a reminder for myself to look over it and see whether I agree with the reasoning and fixes, since both the feature set and performance goals of this code were pretty precise (needed to be ready to create connections easily, but also resettable for changes in config, adding new users and certs, etc.)